proposals/global-attributes-for-main.md

```diff
@@ -0,0 +1,116 @@
+# Attribute target for Main
+
+## Summary
+
+C# should allow an attribute to target the main method / program entrypoint
+without the need for it to be syntactically placed next to it.
+
+## Motivation
+
+Today, top level (simple) programs have no way to place attributes on the main
+method, as there is no syntactic place for them to attach to. This means that
+application models, such as Windows Forms, cannot take advantage of the
+simplified entry point as they require additional attributes to be placed on
+main (see
+<https://github.com/dotnet/designs/blob/main/accepted/2021/winforms/streamline-application-bootstrap.md>).
+
+Further if a source generator wishes to add an attribute to the main method, the
+user of the generator must not use top level statements and explicitly make
+their main method `partial`.
+
+## Detailed design
+
+A new global attribute target `main` will be recognized alongside `assembly` and
+`module`, that allows an attribute to specify it should be attached to the
+entrypoint of the program.
+
+For example:
+
+```csharp
+[main: STAThread]
+```
+
+Can be placed in any syntax tree of the compilation, and the `[STAThread]`
+attribute will be attached the entrypoint of the program.
+
+Both top level statements and regular entrypoints are supported by the attribute
+target. In the case of multiple entrypoints the attribute will only apply to the
+first chosen location.
+
+### Assemblies with no Entry Point
+
+It is an error to use the `main` target location for an attribute in an assembly
+with no entry point (e.g. a class library).
+
+### Attribute locations
+
+There is no new
+[`AttributeTargets`](https://docs.microsoft.com/en-us/dotnet/api/system.attributetargets?view=net-6.0)
+enum value defined as part of this proposal, and it is an error to prefix an
+attribute with `[main:` that has any target other than `All` or `Method`.
```

Comment:
The wording here implies that `AttributeTargets.Property | AttributeTargets.Method` cannot be put in `main:`. Consider rewording to state that the attribute must have a `Method` or `All` target, rather than talking about the things the attribute _doesn't_ have.

Side note: what does a missing `AttributeTargets` translate to? `All`?

Reply:
Good point. Yes, missing AttributeTargets implies `All`.